File: backend/src/test/java/com/back/fairytale/domain/fairytale/controller/FairytaleControllerTest.java
package com.back.fairytale.domain.fairytale.controller;

import com.back.fairytale.domain.fairytale.dto.FairytaleDetailResponse;
import com.back.fairytale.domain.fairytale.dto.FairytaleListResponse;
import com.back.fairytale.domain.fairytale.dto.FairytaleResponse;
import com.back.fairytale.domain.fairytale.entity.Fairytale;
import com.back.fairytale.domain.fairytale.entity.FairytaleKeyword;
import com.back.fairytale.domain.fairytale.repository.FairytaleRepository;
import com.back.fairytale.domain.keyword.entity.Keyword;
import com.back.fairytale.domain.keyword.enums.KeywordType;
import com.back.fairytale.domain.keyword.repository.KeywordRepository;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;

@ActiveProfiles("test")
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc(addFilters = false)
@Transactional
public class FairytaleControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private FairytaleRepository fairytaleRepository;

    @Autowired
    private KeywordRepository keywordRepository;

    // 동화 생성 메서드
    private FairytaleResponse createFairytale(String requestJson) throws Exception {
        MvcResult result = mockMvc.perform(post("/fairytales")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(requestJson))
                .andReturn();

        assertThat(result.getResponse().getStatus()).isEqualTo(200);

        String responseBody = result.getResponse().getContentAsString();
        return objectMapper.readValue(responseBody, FairytaleResponse.class);
    }

    // 동화 생성만 하고 응답이 필요 없는 경우 메서드
    private void createFairytaleOnly(String requestJson) throws Exception {
        MvcResult result = mockMvc.perform(post("/fairytales")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(requestJson))
                .andReturn();

        assertThat(result.getResponse().getStatus()).isEqualTo(200);
    }

    @Test
    @DisplayName("Gemini API를 통한 동화 생성 - 실제 API 호출")
    void t1() throws Exception {
        String requestJson = """
            {
                "childName": "지민",
                "childRole": "용감한 기사",
                "characters": "공주, 마법사",
                "place": "마법의 성",
                "lesson": "용기",
                "mood": "모험적인"
            }
            """;

        FairytaleResponse response = createFairytale(requestJson);

        assertThat(response.id()).isNotNull();
        assertThat(response.title()).isNotNull();
        assertThat(response.childName()).isEqualTo("지민");
        assertThat(response.childRole()).isEqualTo("용감한 기사");

        System.out.println("동화 생성 성공");
    }

    @Test
    @DisplayName("Validation 오류 테스트")
    void t2() throws Exception {
        // childName이 빈 값 (validation 실패)
        String invalidRequest = """
        {
            "childName": "",
            "childRole": "기사",
            "characters": "공주",
            "place": "성",
            "lesson": "용기",
            "mood": "모험"
        }
        """;

        MvcResult result = mockMvc.perform(post("/fairytales")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(invalidRequest))
                .andDo(print())
                .andReturn();

        assertThat(result.getResponse().getStatus()).isEqualTo(400);
    }

    @Test
    @DisplayName("동화 전체 리스트 조회 - 성공")
    void t3() throws Exception {
        String requestJson = """
            {
                "childName": "지민",
                "childRole": "용감한 기사",
                "characters": "공주, 마법사",
                "place": "마법의 성",
                "lesson": "용기",
                "mood": "모험적인"
            }
            """;

        createFairytaleOnly(requestJson);

        MvcResult listResult = mockMvc.perform(get("/fairytales")
                        .contentType(MediaType.APPLICATION_JSON))
                .andDo(print())
                .andReturn();

        assertThat(listResult.getResponse().getStatus()).isEqualTo(200);

        String responseBody = listResult.getResponse().getContentAsString();
        List<FairytaleListResponse> responseList = objectMapper.readValue(
                responseBody, new TypeReference<List<FairytaleListResponse>>() {}
        );

        assertThat(responseList).isNotEmpty();
        assertThat(responseList).hasSize(1);
        assertThat(responseList.get(0).id()).isNotNull();
        assertThat(responseList.get(0).title()).isNotNull();
        assertThat(responseList.get(0).createdAt()).isNotNull();

        System.out.println("사용자별 동화 전체 리스트 조회 성공");
    }

    @Test
    @DisplayName("동화 전체 리스트 조회 - 데이터 없음")
    void t4() throws Exception {
        MvcResult result = mockMvc.perform(get("/fairytales")
                        .contentType(MediaType.APPLICATION_JSON))
                .andDo(print())
                .andReturn();

        assertThat(result.getResponse().getStatus()).isEqualTo(404);
        assertThat(result.getResponse().getContentAsString()).isEqualTo("등록된 동화가 없습니다.");
    }

    @Test
    @DisplayName("동화 생성 시 키워드 저장 검증")
    void t5() throws Exception {
        String requestJson = """
            {
                "childName": "민수",
                "childRole": "마법사",
                "characters": "용, 공주, 왕자",
                "place": "마법의 숲, 높은 탑",
                "lesson": "용기와 지혜",
                "mood": "신비로운"
            }
            """;

        // 동화 생성
        FairytaleResponse response = createFairytale(requestJson);

        Optional<Fairytale> savedFairytale = fairytaleRepository.findById(response.id());
        assertThat(savedFairytale).isPresent();

        Fairytale fairytale = savedFairytale.get();

        // FairytaleKeyword를 통한 키워드 검증
        List<FairytaleKeyword> fairytaleKeywords = fairytale.getFairytaleKeywords();
        assertThat(fairytaleKeywords).isNotEmpty();

        assertThat(fairytaleKeywords)
                .extracting(fk -> fk.getKeyword().getKeywordType())
                .contains(
                        KeywordType.CHILD_NAME,
                        KeywordType.CHILD_ROLE,
                        KeywordType.CHARACTERS,
                        KeywordType.PLACE,
                        KeywordType.LESSON,
                        KeywordType.MOOD
                );

        // 아이이름 키워드 검증
        Optional<FairytaleKeyword> childNameKeyword = fairytaleKeywords.stream()
                .filter(fk -> fk.getKeyword().getKeywordType() == KeywordType.CHILD_NAME)
                .findFirst();
        assertThat(childNameKeyword).isPresent();
        assertThat(childNameKeyword.get().getKeyword().getKeyword()).isEqualTo("민수");

        // 등장인물 키워드 검증
        List<FairytaleKeyword> characterKeywords = fairytaleKeywords.stream()
                .filter(fk -> fk.getKeyword().getKeywordType() == KeywordType.CHARACTERS)
                .toList();
        assertThat(characterKeywords).hasSize(3);
        assertThat(characterKeywords)
                .extracting(fk -> fk.getKeyword().getKeyword())
                .containsExactlyInAnyOrder("용", "공주", "왕자");

        // 교훈 키워드 검증
        List<FairytaleKeyword> lessonKeywords = fairytaleKeywords.stream()
                .filter(fk -> fk.getKeyword().getKeywordType() == KeywordType.LESSON)
                .toList();
        assertThat(lessonKeywords).hasSize(1);
        assertThat(lessonKeywords.get(0).getKeyword().getKeyword()).isEqualTo("용기와 지혜");
    }

    @Test
    @DisplayName("같은 키워드 재사용 검증 - 중복 키워드는 새로 생성하지 않음")
    void t6() throws Exception {
        // 첫번째 동화 생성
        String firstRequestJson = """
            {
                "childName": "지민",
                "childRole": "기사",
                "characters": "공주",
                "place": "성",
                "lesson": "용기",
                "mood": "모험적인"
            }
            """;

        createFairytaleOnly(firstRequestJson);

        // 첫 번째 동화 생성 후 키워드 개수 확인
        long keywordCountAfterFirst = keywordRepository.count();

        // 동일한 키워드를 포함한 두 번째 동화 생성
        String secondRequestJson = """
            {
                "childName": "민수",
                "childRole": "기사",
                "characters": "공주",
                "place": "성",
                "lesson": "용기",
                "mood": "신비로운"
            }
            """;

        createFairytaleOnly(secondRequestJson);

        long keywordCountAfterSecond = keywordRepository.count();

        // 키워드 4개 중복되고 새로운 키워드는 2개니까 총 2개의 키워드만 추가되어야 함
        assertThat(keywordCountAfterSecond).isEqualTo(keywordCountAfterFirst + 2);

        // 전체에서 해당 키워드 개수 세기 -> 중복 저장인지 확인
        List<Keyword> allKeywords = keywordRepository.findAll();
        long countKeyword1 = allKeywords.stream()
                .filter(k -> k.getKeyword().equals("기사") && k.getKeywordType() == KeywordType.CHILD_ROLE)
                .count();
        assertThat(countKeyword1).isEqualTo(1); // 하나만 존재 (중복 저장 안되니까)

        long countKeyword2 = allKeywords.stream()
                .filter(k -> k.getKeyword().equals("공주") && k.getKeywordType() == KeywordType.CHARACTERS)
                .count();
        assertThat(countKeyword2).isEqualTo(1);
    }

    @Test
    @DisplayName("사용자별 동화 상세 조회 - 성공")
    void t7() throws Exception {
        String requestJson = """
        {
            "childName": "하늘",
            "childRole": "마법사",
            "characters": "용, 요정",
            "place": "마법의 숲",
            "lesson": "용기",
            "mood": "환상적인"
        }
        """;

        FairytaleResponse response = createFairytale(requestJson);
        Long fairytaleId = response.id();

        MvcResult detailResult = mockMvc.perform(get("/fairytales/" + fairytaleId)
                        .contentType(MediaType.APPLICATION_JSON))
                .andDo(print())
                .andReturn();

        assertThat(detailResult.getResponse().getStatus()).isEqualTo(200);

        String detailResponseBody = detailResult.getResponse().getContentAsString();
        FairytaleDetailResponse detailResponse = objectMapper.readValue(detailResponseBody, FairytaleDetailResponse.class);

        assertThat(detailResponse.id()).isEqualTo(fairytaleId);
        assertThat(detailResponse.title()).isNotNull();
        assertThat(detailResponse.content()).isNotNull();
        assertThat(detailResponse.childName()).isEqualTo("하늘");
        assertThat(detailResponse.childRole()).isEqualTo("마법사");
        assertThat(detailResponse.characters()).contains("용", "요정");
        assertThat(detailResponse.place()).isEqualTo("마법의 숲");
        assertThat(detailResponse.lesson()).isEqualTo("용기");
        assertThat(detailResponse.mood()).isEqualTo("환상적인");
        assertThat(detailResponse.createdAt()).isNotNull();

        System.out.println("사용자별 동화 상세 조회 성공");
    }

    @Test
    @DisplayName("동화 삭제 - 성공")
    void t10() throws Exception {
        String requestJson = """
        {
            "childName": "하늘",
            "childRole": "마법사",
            "characters": "용, 요정",
            "place": "마법의 숲",
            "lesson": "용기",
            "mood": "환상적인"
        }
        """;

        FairytaleResponse response = createFairytale(requestJson);
        Long fairytaleId = response.id();

        MvcResult deleteResult = mockMvc.perform(delete("/fairytales/" + fairytaleId)
                        .contentType(MediaType.APPLICATION_JSON))
                .andDo(print())
                .andReturn();

        // 삭제 성공 확인
        assertThat(deleteResult.getResponse().getStatus()).isEqualTo(204);

        // 삭제 후 존재하지 않는지 확인
        assertThat(fairytaleRepository.findById(fairytaleId)).isEmpty();

        System.out.println("동화 삭제 성공");
    }

    @Test
    @DisplayName("동화 생성과 이미지 생성 통합 테스트")
    void t11() throws Exception {
        String requestJson = """
        {
            "childName": "소라",
            "childRole": "바다 공주",
            "characters": "돌고래, 거북이",
            "place": "바다",
            "lesson": "친구의 소중함",
            "mood": "따듯한, 신비로운"
        }
        """;

        FairytaleResponse response = createFairytale(requestJson);

        // 동화 생성 확인
        assertThat(response.id()).isNotNull();
        assertThat(response.title()).isNotNull();
        assertThat(response.content()).isNotNull();
        assertThat(response.childName()).isEqualTo("소라");

        // 이미지 URL 확인 (현재는 null)
        System.out.println("생성된 동화 ID: " + response.id());
        System.out.println("이미지 URL: " + response.imageUrl()); // 현재는 null

        System.out.println("통합 테스트 - 동화 생성 성공!");
    }
}

File: backend/src/test/java/com/back/fairytale/domain/fairytale/controller/FairytaleControllerTest.kt
package com.back.fairytale.domain.fairytale.controller

import com.back.fairytale.domain.fairytale.dto.FairytaleCreateRequest
import com.back.fairytale.domain.fairytale.dto.FairytaleDetailResponse
import com.back.fairytale.domain.fairytale.dto.FairytaleResponse
import com.back.fairytale.domain.fairytale.entity.Fairytale
import com.back.fairytale.domain.fairytale.repository.FairytaleRepository
import com.back.fairytale.domain.keyword.entity.Keyword
import com.back.fairytale.domain.keyword.enums.KeywordType
import com.back.fairytale.domain.keyword.repository.KeywordRepository
import com.back.fairytale.domain.user.entity.User
import com.back.fairytale.domain.user.repository.UserRepository
import com.back.fairytale.external.ai.client.GeminiClient
import com.back.fairytale.external.ai.client.HuggingFaceClient
import com.back.fairytale.global.security.CustomOAuth2User
import com.back.fairytale.global.util.impl.GoogleCloudStorage
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.cloud.storage.Storage
import com.ninjasquad.springmockk.MockkBean
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockMvc(addFilters = false)
@Transactional
class FairytaleControllerTest {

    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var fairytaleRepository: FairytaleRepository

    @Autowired
    lateinit var keywordRepository: KeywordRepository

    @MockkBean lateinit var geminiClient: GeminiClient
    @MockkBean lateinit var huggingFaceClient: HuggingFaceClient
    @MockkBean(relaxed = true) lateinit var googleCloudStorage: GoogleCloudStorage
    @MockkBean lateinit var storage: Storage

    private var savedUserId: Long = -1L

    @BeforeEach
    fun setUpUser() {
        savedUserId = userRepository.save(
            User(
                name = "tester",
                nickname = "nick",
                email = "tester@example.com",
                socialId = "social"
            )
        ).id!!

        // SecurityContext에 인증 객체 직접 세팅
        val principal = CustomOAuth2User(savedUserId, "tester", "ROLE_USER")
        val auth = UsernamePasswordAuthenticationToken(
            principal,
            null,
            listOf(SimpleGrantedAuthority("ROLE_USER"))
        )
        SecurityContextHolder.getContext().authentication = auth
    }

    @Test
    @DisplayName("동화 생성 성공")
    fun t1() {
        val req = FairytaleCreateRequest(
            childName = "민지",
            childRole = "모험가",
            characters = "토끼, 거북이",
            place = "숲속 마을",
            lesson = "협동, 우정",
            mood = "따뜻한, 유쾌한"
        )

        // Gemini 스텁
        every { geminiClient.generateFairytale(match { it.contains("응답 형식: [제목:") }) } returns """
            [제목: 민지의 용감한 모험]
            민지는 숲속 마을에서 토끼와 거북이와 함께 협동의 가치를 배우는 모험을 떠났어요...
        """.trimIndent()
        every { geminiClient.generateFairytale(match { it.contains("영어로 번역") }) } returns
                "a cute child adventurer with a rabbit and a turtle, smiling, warm tone, 1:1 illustration"

        // HuggingFace/GCS 스텁
        every { huggingFaceClient.generateImage(any()) } returns "fake-image".toByteArray()
        every { googleCloudStorage.uploadImageBytesToCloud(any()) } returns
                "https://storage.googleapis.com/bucket/fairy.png"

        val mvcResult = mockMvc.perform(
            post("/fairytales")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
        )
            .andExpect(status().isOk)
            .andReturn()

        val body = mvcResult.response.contentAsString
        val res = objectMapper.readValue(body, FairytaleResponse::class.java)

        assertThat(res.id).isPositive()
        assertThat(res.title).isEqualTo("민지의 용감한 모험")
        assertThat(res.content).contains("민지", "숲속 마을", "협동")
        assertThat(res.imageUrl).isEqualTo("https://storage.googleapis.com/bucket/fairy.png")
        assertThat(res.childName).isEqualTo("민지")
        assertThat(res.childRole).isEqualTo("모험가")
        assertThat(res.characters).isEqualTo("토끼, 거북이")
        assertThat(res.place).isEqualTo("숲속 마을")
        assertThat(res.lesson).isEqualTo("협동, 우정")
        assertThat(res.mood).isEqualTo("따뜻한, 유쾌한")
        assertThat(res.userId).isEqualTo(savedUserId)
        assertThat(res.createdAt).isNotNull()
    }

    @Test
    @DisplayName("동화 생성 실패 - childName 공백(NotBlank 위반)")
    fun t2() {
        val req = FairytaleCreateRequest(
            childName = "   ",
            childRole = "모험가",
            characters = "토끼, 거북이",
            place = "숲속 마을",
            lesson = "협동, 우정",
            mood = "따뜻한, 유쾌한"
        )

        val mvcResult = mockMvc.perform(
            post("/fairytales")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(req))
        )
            .andExpect(status().isBadRequest)
            .andReturn()

        // 외부 의존 호출이 전혀 없어야 함
        verify(exactly = 0) { geminiClient.generateFairytale(any()) }
        verify(exactly = 0) { huggingFaceClient.generateImage(any()) }
        verify(exactly = 0) { googleCloudStorage.uploadImageBytesToCloud(any()) }
        confirmVerified(geminiClient, huggingFaceClient, googleCloudStorage)
    }

    @Test
    @DisplayName("내 동화 전체 조회 성공")
    fun t3() {
        val user = userRepository.findById(savedUserId).get()

        // 동화 2개를 시간 차이를 두고 저장
        val f1 = fairytaleRepository.save(
            Fairytale(
                user = user,
                title = "첫 번째 동화",
                content = "내용1",
                imageUrl = null,
                isPublic = false
            )
        )
        Thread.sleep(10)
        val f2 = fairytaleRepository.save(
            Fairytale(
                user = user,
                title = "두 번째 동화",
                content = "내용2",
                imageUrl = null,
                isPublic = true
            )
        )

        val mvcResult = mockMvc.perform(
            get("/fairytales")
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(status().isOk)
            .andReturn()

        val body = mvcResult.response.contentAsString
        val res = objectMapper.readTree(body)

        assertThat(res.isArray).isTrue()
        assertThat(res.size()).isEqualTo(2)

        assertThat(res[0].get("title").asText()).isEqualTo("두 번째 동화")
        assertThat(res[1].get("title").asText()).isEqualTo("첫 번째 동화")
    }

    @Test
    @DisplayName("동화 상세 조회 성공")
    fun t4() {
        val user = userRepository.findById(savedUserId).get()

        // 동화 상세 조회는 키워드 필요
        val k1 = keywordRepository.save(Keyword.of("민지", KeywordType.CHILD_NAME))
        val k2 = keywordRepository.save(Keyword.of("모험가", KeywordType.CHILD_ROLE))
        val k3 = keywordRepository.save(Keyword.of("토끼", KeywordType.CHARACTERS))
        val k4 = keywordRepository.save(Keyword.of("거북이", KeywordType.CHARACTERS))
        val k5 = keywordRepository.save(Keyword.of("숲속 마을", KeywordType.PLACE))
        val k6 = keywordRepository.save(Keyword.of("협동, 우정", KeywordType.LESSON))
        val k7 = keywordRepository.save(Keyword.of("따뜻한, 유쾌한", KeywordType.MOOD))

        val fairy = Fairytale(
            user = user,
            title = "상세 테스트 동화",
            content = "내용 상세",
            imageUrl = "https://storage.googleapis.com/bucket/detail.png",
            isPublic = false
        )

        listOf(k1, k2, k3, k4, k5, k6, k7).forEach { fairy.addKeyword(it) }
        fairytaleRepository.save(fairy)

        val mvcResult = mockMvc.perform(
            get("/fairytales/{id}", fairy.id!!)
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(status().isOk)
            .andReturn()

        val body = mvcResult.response.contentAsString
        val res = objectMapper.readValue(body, FairytaleDetailResponse::class.java)

        assertThat(res.childName).isEqualTo("민지")
        assertThat(res.place).isEqualTo("숲속 마을")
        assertThat(res.id).isEqualTo(fairy.id)
        assertThat(res.title).isEqualTo("상세 테스트 동화")
        assertThat(res.content).isEqualTo("내용 상세")
        assertThat(res.imageUrl).isEqualTo("https://storage.googleapis.com/bucket/detail.png")
        assertThat(res.isPublic).isFalse()
        assertThat(res.createdAt).isNotNull()
    }

    @Test
    @DisplayName("동화 상세 조회 실패 - 존재하지 않는 ID")
    fun t5() {
        val notExistId = 999L

        val mvcResult = mockMvc.perform(
            get("/fairytales/{id}", notExistId)
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(status().isNotFound)
            .andReturn()

        val body = mvcResult.response.contentAsString
        assertThat(body).contains("동화를 찾을 수") // 서비스 메시지 일부만 느슨히 검증
    }

    @Test
    @DisplayName("동화 삭제 성공")
    fun t6() {
        val user = userRepository.findById(savedUserId).get()
        val imageUrl = "https://storage.googleapis.com/bucket/to-delete.png"
        val saved = fairytaleRepository.save(
            Fairytale(
                user = user,
                title = "삭제 대상 동화",
                content = "삭제 컨텐츠",
                imageUrl = imageUrl,
                isPublic = false
            )
        )

        mockMvc.perform(
            delete("/fairytales/{id}", saved.id!!)
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(status().isNoContent)

        // DB도 검증
        val exists = fairytaleRepository.findById(saved.id!!).isPresent
        assertThat(exists).isFalse()

        // 이미지 삭제 호출되었는지
        verify(exactly = 1) { googleCloudStorage.deleteImages(match { it == listOf(imageUrl) }) }
    }

}
